### PR TITLE
add initializer design

### PR DESCRIPTION
**Pull request by hongchaodeng**
#1787 #999

### CODE REVIEW

**Review by wonderflow — COMMENTED** [review]
LGTM, please add Implementation Plan

**Comment by wonderflow** [comment]
I have offline discussed with @hongchaodeng 

1. we should still create a Env Object for user convenience, so users can clearly understand the concept.
2. env should be able to refer to another env for dependency and reuse.

File: design/vela-core/environment.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  name: prod-env
+spec:
+  components:
+    - name: prod-cluster-env
+      type: cluster-env
+      properties:
+        # cluster with given names or matched labels will belong to this env (i.e. prod).
+        clusterSelector:
+          names: ["prod-1"， "prod-2"]
+          labels:
```
Comment (zzxwill) [inline design/vela-core/environment.md:51]: 看到你下面解释了 `cluster-env`，可以再举一个 cluster-env 的 ComponentDefinition 的 yaml 范例吗？这里 properties 也是两层结构，对用户不太友好，也可能是我没有完全理解 cluster-env 的含义。

File: design/vela-core/environment.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
+# Managing Environments As Shared Bases For Application Deployment
+
+## Background
+
+A team of application development usually needs to prepare some shared environments for developers to deploy applications to.
+For example, most teams will have two environments, a *dev* environment for testing applications,
```
Comment (zzxwill) [inline design/vela-core/environment.md:6]:
```suggestion
For example, most teams will have two environments, a *staging* environment for testing applications,
```
Normally we call it testing or staging environment.

Comment (hongchaodeng) [reply design/vela-core/environment.md:6]: I would like to keep it as dev which is simpler.

File: design/vela-core/environment.md (hunk 1 above)
Comment (hongchaodeng) [reply design/vela-core/environment.md:51]: Yeah. There will be more details for the definition in implementation plan. THe current section is for reviewing high-level design.

**Comment by hongchaodeng** [comment]
> I have offline discussed with @hongchaodeng
> we should still create a Env Object for user convenience, so users can clearly understand the concept.
> env should be able to refer to another env for dependency and reuse.

@wonderflow Updated design.

File: design/vela-core/environment.md [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+We propose to add an Environment CRD:
+
+```yaml
+kind: Environment
+metadata:
+  name: prod
+spec:
+  # An environment consists of multiple system components.
+  # An environment is a grouping of these shared resources for user applications.
+  components:
```
Comment (wonderflow) [inline design/vela-core/environment.md:45]: let's name it resources? And we can create K8s namespace for Environment here.

Comment (wonderflow) [reply design/vela-core/environment.md:45]: Also, we can create scopes(SLO, VPC, Identidy) here.

File: design/vela-core/environment.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+```yaml
+kind: Environment
+metadata:
+  name: prod
+spec:
+  # An environment consists of multiple system resources.
+  # An environment is a grouping of these shared resources for user applications.
+  resources:
+    - template:
```
Comment (wonderflow) [inline design/vela-core/environment.md:46]: why call it template?

We can category and distinguish here:

* `system`:  install only once, if the system already has one, no need to install again.
* `env`: install per env, always install, and will report an error if already exist.

Comment (hongchaodeng) [reply design/vela-core/environment.md:46]: > env: install per env, always install, and will report an error if already exist.

I didn't catch this. Can you give an example?

Comment (hongchaodeng) [reply design/vela-core/environment.md:46]: Note: we will add a field to indicate that this resource should be **owned** by this env.

**Review by wonderflow — COMMENTED** [review]
the design LGTM, please add implementation plan.

File: design/vela-core/environment.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        spec:
+          components:
+            - name: prod-env-clusters
+              type: cluster-selector
+              properties:
+                # cluster with given names or matched labels will belong to this env
+                clusterSelector:
+                  names: ["prod-1"， "prod-2"]
+                  labels:
+                    tier: prod
```
Comment (wonderflow) [inline design/vela-core/environment.md:107]: 这部分我想了下要提到跟 resources 平级，不要跟应该部署的资源混在一起

Comment (hongchaodeng) [reply design/vela-core/environment.md:107]: 详细写一下你的想法，包括：你想到的用户场景和问题是什么？应该怎么做？为什么要这么做？

Comment (wonderflow) [reply design/vela-core/environment.md:107]: 用户指定一个环境对应某些集群，跟一个环境下需要哪些系统组件，是两个维度的事情。

Comment (hongchaodeng) [reply design/vela-core/environment.md:107]: 用不同 components 来表示即可

Comment (wonderflow) [reply design/vela-core/environment.md:107]: 集群的信息是动态变化的，需要实际应用部署的时候每次去计算（利用ocm这样的组件的话就是每次都要生成placement），应该是一个metadata的标记，而不是一个Component

Comment (wonderflow) [reply design/vela-core/environment.md:107]: 我们对Component的定义不能泛化，只是metadata的记录，而且是可枚举的情况，应该提到外面来。

Comment (hongchaodeng) [reply design/vela-core/environment.md:107]: > 集群的信息是动态变化的，需要实际应用部署的时候每次去计算（利用ocm这样的组件的话就是每次都要生成placement），应该是一个metadata的标记，而不是一个Component

这里并没有限制集群的信息不能动态变化，不是吗？所以才有 label selector 这样的机制去动态选择。

> 我们对Component的定义不能泛化，只是metadata的记录，而且是可枚举的情况，应该提到外面来。

没看懂这句话。能具体说说有什么问题？

**Comment by hongchaodeng** [comment]
Add implementation plan for environment controller first. Let's implement it first and add plans for other operators later.

File: design/vela-core/environment.md [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+We propose to add an Environment CRD:
+
+```yaml
+kind: Environment
+metadata:
+  name: prod
+spec:
+  # An environment consists of multiple system resources.
+  # An environment is a grouping of these shared resources for user applications.
+  resources:
```
Comment (wonderflow) [inline design/vela-core/environment.md:45]:
```suggestion
  initializer:
```

File: design/vela-core/environment.md [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+- Platform team defines user Components for Applications.
+- Developers choose user Components and Environments to deploy Applicatons.
+
+When deploying environments, there are additional requirements:
+
+- The ability to describe dependency relations between environments. For example, both *dev* and *prod* environments
+  would depend on a common environment that sets up operators both *dev* and *prod*  would use.
+
+
+## Environment CRD
```
Comment (wonderflow) [inline design/vela-core/environment.md:34]:
```suggestion
## Initializer CRD
```

File: design/vela-core/environment.md [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+We propose to add an Environment CRD:
+
+```yaml
+kind: Environment
```
Comment (wonderflow) [inline design/vela-core/environment.md:39]:
```suggestion
kind: Initializer
```

File: design/vela-core/environment.md [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+        kind: Application
+        spec:
+          ...
+
+  # dependsOn indicates the other environments that this env depends on.
+  # The environment will not apply its components until all dependencies exist.
+  dependsOn:
+    - ref:
+        apiVersion: core.oam.dev/v1beta1
+        kind: Environment
```
Comment (wonderflow) [inline design/vela-core/environment.md:63]:
```suggestion
        kind: Initializer
```

File: design/vela-core/environment.md [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+  resources:
+    - template:
+        apiVersion: core.oam.dev/v1beta1
+        kind: Application
+        metadata:
+          name: prod-env-clusters
+        spec:
+          components:
+            - name: prod-env-clusters
+              type: cluster-selector
```
Comment (wonderflow) [inline design/vela-core/environment.md:101]: 这个是initializer的placement rule，表示这个initializer要部署的位置，跟initializer里面要不是的资源不能混在一起

File: design/vela-core/environment.md [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+- Add a new Environment CRD and controller skeleton in KubeVela.
+- In reconcile logic:
+  - Check environment dependency
+  - Handle each resource:
+    - Render the template into a resource object.
+    - If `mustOwn` is true, check if the resource ownerRef is the current Environment CR.
+    - Put Vela-related labels into the object.
+    - Put current Environment CR into the ownerRef of the object.
+    - Apply the resource object.
+
```
Comment (wonderflow) [inline design/vela-core/environment.md:306]: initializer的删除应该要能删除对应自己own的资源，这个如何实现，使用ownerReference是否OK？

**Comment by hongchaodeng** [comment]
Updated:
- rename env -> initializer
- change template -> component
- add finalizer implementation detail

**Comment by hongchaodeng** [comment]
Updated:

- Change components -> appTemplate
- Use policy to describe placement, slo-policy, chaos-policy.

**Review by wonderflow — APPROVED** [review]

**Pull request merged**